daft/iceberg/iceberg_write.py

```diff
@@ -0,0 +1,180 @@
+        from pyiceberg.transforms import (
+            BucketTransform,
+            DayTransform,
+            HourTransform,
+            IdentityTransform,
+            MonthTransform,
+            TruncateTransform,
+            YearTransform,
+        )
+
+        series = Series.from_arrow(array)
+
+        transformed = None
+        if isinstance(transform, IdentityTransform):
+            transformed = series
+        elif isinstance(transform, YearTransform):
+            transformed = series.partitioning.years()
+        elif isinstance(transform, MonthTransform):
+            transformed = series.partitioning.months()
+        elif isinstance(transform, DayTransform):
+            transformed = series.partitioning.days()
+        elif isinstance(transform, HourTransform):
+            transformed = series.partitioning.hours()
+        elif isinstance(transform, BucketTransform):
+            n = transform.num_buckets
+            transformed = series.partitioning.iceberg_bucket(n)
+        elif isinstance(transform, TruncateTransform):
+            w = transform.width
+            transformed = series.partitioning.iceberg_truncate(w)
+        else:
+            warnings.warn(f"{transform} not implemented, Please make an issue!")
+
+        return transformed.to_arrow() if transformed is not None else None
+
+    partition_values_table = pa.table(
+        {
+            str(partition.field_id): partition_transform(arrow_table[field.name], partition.transform)
+            for partition, field in partition_columns
+        }
+    )
+
+    # Sort by partitions
+    sort_indices = pa.compute.sort_indices(
+        partition_values_table,
+        sort_keys=[(col, "ascending") for col in partition_values_table.column_names],
+        null_placement="at_end",
+    ).to_pylist()
+    arrow_table = arrow_table.take(sort_indices)
+
+    # Get slice_instructions to group by partitions
+    partition_values_table = partition_values_table.take(sort_indices)
+    reversed_indices = pa.compute.sort_indices(
+        partition_values_table,
+        sort_keys=[(col, "descending") for col in partition_values_table.column_names],
+        null_placement="at_start",
+    ).to_pylist()
+    slice_instructions: List[Dict[str, Any]] = []
+    last = len(reversed_indices)
+    reversed_indices_size = len(reversed_indices)
+    ptr = 0
+    while ptr < reversed_indices_size:
+        group_size = last - reversed_indices[ptr]
+        offset = reversed_indices[ptr]
+        slice_instructions.append({"offset": offset, "length": group_size})
+        last = reversed_indices[ptr]
+        ptr = ptr + group_size
+
+    table_partitions: List[_TablePartition] = _get_table_partitions(arrow_table, spec, schema, slice_instructions)
+
+    return table_partitions
+
+
+def micropartition_to_arrow_tables(
+    table: MicroPartition, path: str, schema: "IcebergSchema", partition_spec: "IcebergPartitionSpec"
+) -> List[Tuple["pa.Table", str, "Record"]]:
```

Comment:
The preferred data representation across language/system boundaries is usually an iterator of RecordBatches. Is it possible to use that here?

Noting also that `pa.Table` is a pyarrow-only concept, and not actually in the Arrow spec

Reply:
I'm returning a `pa.Table` here since that is what `_write_tabular_arrow_table` in table_io.py takes in